### PR TITLE
fix: copy git tracked files (#912)

### PR DESCRIPTION
**Pull request by jsoref**
So, this is almost certainly not the right fix for #912, but for git oriented projects it does something which would probably be good enough for me anyway. -- Note: I'm not a go dev, so the style here is probably very wrong, it's just an attempt at playing with things. I'm sure I could find the proper go-git function to ask "are you tracked", but I decided to start with the simpler "go ls-files" approach.

# Before

```
% act
[What do I have/Introspection] 🚀  Start image=catthehacker/ubuntu:act-latest
[What do I have/Introspection]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=false
[What do I have/Introspection]   🐳  docker create image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[What do I have/Introspection]   🐳  docker run image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[What do I have/Introspection]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[What do I have/Introspection]   🐳  docker cp src=/Users/jsoref/code/nektos/ignored/. dst=/Users/jsoref/code/nektos/ignored
[What do I have/Introspection]   🐳  docker exec cmd=[mkdir -p /Users/jsoref/code/nektos/ignored] user= workdir=
[What do I have/Introspection] ⭐  Run checkout
[What do I have/Introspection]   ✅  Success - checkout
[What do I have/Introspection] ⭐  Run introspect
[What do I have/Introspection]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jsoref/code/nektos/ignored/workflow/1] user= workdir=
| total 4
| drwxr-xr-x 2 root root 4096 Jan  2 06:15 workflow
[What do I have/Introspection]   ✅  Success - introspect
```

# After
```
% ../act/dist/local/act
[What do I have/Introspection] 🚀  Start image=catthehacker/ubuntu:act-latest
[What do I have/Introspection]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=false
[What do I have/Introspection]   🐳  docker create image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[What do I have/Introspection]   🐳  docker run image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[What do I have/Introspection]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[What do I have/Introspection]   🐳  docker cp src=/Users/jsoref/code/nektos/ignored/. dst=/Users/jsoref/code/nektos/ignored
[What do I have/Introspection]   🐳  docker exec cmd=[mkdir -p /Users/jsoref/code/nektos/ignored] user= workdir=
[What do I have/Introspection] ⭐  Run checkout
[What do I have/Introspection]   ✅  Success - checkout
[What do I have/Introspection] ⭐  Run introspect
[What do I have/Introspection]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/1] user= workdir=
| total 0
| -rw-r--r-- 1 502 dialout 0 Jan  2 05:49 README.md
[What do I have/Introspection]   ✅  Success - introspect
```

### CODE REVIEW

**Comment by KnisterPeter** [comment]
First of all thank you :smiley: 

I find it hard to review currently, as we don't have tests (at least not much) for that.
I think it might really be good to separate your feature into it's own file with a function which decides if a file should be copied or not.
Then it would be quite easy to add test for that.

File: pkg/container/docker_run.go [hunk 1]
```diff
@@ -576,34 +578,11 @@ func (cr *containerReference) copyDir(dstPath string, srcPath string, useGitIgno
...
-				}
-				return nil
-			}
-
+		addEntry := func(file string, sansPrefix string, fi os.FileInfo) error {
 			// return on non-regular files (thanks to [kumo](https://medium.com/@komuw/just-like-you-did-fbdd7df829d3) for this suggested update)
 			linkName := fi.Name()
 			if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
-				linkName, err = os.Readlink(file)
+				_, err := os.Readlink(file)
```
[the remark this answers is not in the document]
Comment (jsoref) [reply pkg/container/docker_run.go:585]: I definitely wasn't pleased about the linter nor that I couldn't figure out the need.

Will see about the right change.

File: pkg/container/docker_run.go [hunk 2]
```diff
@@ -647,7 +626,77 @@ func (cr *containerReference) copyDir(dstPath string, srcPath string, useGitIgno
...
+								return err
+							}
+						}
+					}
+				}
+			}
+			return err
+		}
+		if useGitIgnore {
+			if tryGitIgnore() != nil {
```
[the remark this answers is not in the document]
Comment (jsoref) [reply pkg/container/docker_run.go:673]: My hope was to abandon what I had, but thinking about it, I suspect I need to actually reset my stash and that I haven't done that yet.

**Comment by jsoref** [comment]
That seems reasonable. If you have any preferences about naming, please feel free to provide it before I start.

I'm not a big go dev, but so far, the ability to refactor the code here makes me fairly confident that moving it into a distinct file wouldn't be a big deal.

**Comment by KnisterPeter** [comment]
> That seems reasonable. If you have any preferences about naming, please feel free to provide it before I start.
> 

I have no preference, just give it a start.

> I'm not a big go dev, but so far, the ability to refactor the code here makes me fairly confident that moving it into a distinct file wouldn't be a big deal.

It should be straight forward if you stay in the folder (same package).

**Comment by jsoref** [comment]
Maybe i'll look again this week...

**Pull request closed without merging**